packages/flutter/lib/src/material/date.dart

```diff
@@ -0,0 +1,225 @@
+    final int weekdayFromMonday = DateTime(year, month).weekday - 1;
+
+    // 0-based start of week depending on the locale, with 0 representing Sunday.
+    int firstDayOfWeekIndex = localizations.firstDayOfWeekIndex;
+
+    // firstDayOfWeekIndex recomputed to be Monday-based, in order to compare with
+    // weekdayFromMonday.
+    firstDayOfWeekIndex = (firstDayOfWeekIndex - 1) % 7;
+
+    // Number of days between the first day of week appearing on the calendar,
+    // and the day corresponding to the first of the month.
+    return (weekdayFromMonday - firstDayOfWeekIndex) % 7;
+  }
+
+  /// Returns the number of days in a month, according to the proleptic
+  /// Gregorian calendar.
+  ///
+  /// This applies the leap year logic introduced by the Gregorian reforms of
+  /// 1582. It will not give valid results for dates prior to that time.
+  static int getDaysInMonth(int year, int month) {
+    if (month == DateTime.february) {
+      final bool isLeapYear = (year % 4 == 0) && (year % 100 != 0) || (year % 400 == 0);
+      return isLeapYear ? 29 : 28;
+    }
+    const List<int> daysInMonth = <int>[31, -1, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+    return daysInMonth[month - 1];
+  }
+}
+
+/// Mode of the date picker dialog.
+///
+/// Either a calendar or text input. In [calendar] mode, a calendar view is
+/// displayed and the user taps the day they wish to select. In [input] mode a
+/// [TextField] is displayed and the user types in the date they wish to select.
+///
+/// See also:
+///
+///  * [showDatePicker] and [showDateRangePicker], which use this to control
+///    the initial entry mode of their dialogs.
+enum DatePickerEntryMode {
+  /// Tapping on a calendar.
+  calendar,
+
+  /// Text input.
+  input,
+}
+
+/// Initial display of a calendar date picker.
+///
+/// Either a grid of available years or a monthly calendar.
+///
+/// See also:
+///
+///  * [showDatePicker], which shows a dialog that contains a material design
+///    date picker.
+///  * [CalendarDatePicker], widget which implements the material design date picker.
+enum DatePickerMode {
+  /// Choosing a month and day.
+  day,
+
+  /// Choosing a year.
+  year,
+}
+
+/// Signature for predicating dates for enabled date selections.
+///
+/// See [showDatePicker], which has a [SelectableDayPredicate] parameter used
+/// to specify allowable days in the date picker.
+typedef SelectableDayPredicate = bool Function(DateTime day);
+
+/// Encapsulates a start and end [DateTime] that represent the range of dates
+/// between them.
+///
+/// See also:
+///  * [showDateRangePicker], which displays a dialog that allows the user to
+///    select a date range.
+@immutable
+class DateTimeRange {
+  /// Creates a date range for the given start and end [DateTime].
+  ///
+  /// [start] and [end] must be non-null.
+  const DateTimeRange({
+    required this.start,
+    required this.end,
+  }) : assert(start != null),
```

Comment:
I assume that we're OK with "negative" data ranges.

Reply:
Actually we aren't, so I added an assert and documented this.